[Wong An Wei] iP

AnWe11 opened the pull request:
### **Luigi**

> “Your mind is for having ideas, not holding them.” – David Allen

Luigi helps to keep track of your tasks so you dont have to! 😉

Its features include,

- Adding / Deleting tasks
- Marking/ Unmarking tasks
- Local memory update
- Encouraging messages from Luigi

All you need to do is, 

1. Download it from [here](https://github.com/AnWe11/ip/releases/tag/Luigi-v2)
2. Double Click the file
3. Add your tasks
4. Let it manage for you

Other features **coming soon** include,

- _[ ] Reminders_
- _[ ] Prioritisation of tasks_

`
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

AnWe11 committed:
Test if SourceTree is set up correctly

AnWe11 committed:
Renamed Duke to Luigi, added Greetings and Goodbye Strings

AnWe11 committed:
Updated Luigi Logo

AnWe11 committed:
Added Tasks class, Set up Chatbot's ability to echo inputs from user

AnWe11 committed:
Added comments to code

AnWe11 committed:
Added list function to display list of current Tasks

AnWe11 committed:
Added mark / unmark function

AnWe11 committed:
Abstracted command keyword class, added a new function that will display remaining tasks whenever user marks / unmarks a task

AnWe11 committed:
Added new tasks, todos, deadlines, and events, abstracted all these and made them into childre of parent class Tasks

AnWe11 committed:
Added exception handling for incorrect input formatting, added delete function but indexing is still broken

AnWe11 committed:
Added delete function with the list updating properly, added more exception handling, added a task manager class

AnWe11 committed:
Added automated text testing, have yet to add specified input and output

AnWe11 committed:
Package everything into respective packages

AnWe11 committed:
Redefine Enums

AnWe11 committed:
Clean up code and formatting

AnWe11 committed:
rename files

AnWe11 committed:
rename packages

AnWe11 committed:
add save, mark/unmark and delete still not fix

AnWe11 committed:
Fix saving and updated the other functions accordingly

AnWe11 committed:
Clean up delete function to sync with storage

AnWe11 committed:
Merge branch 'branch-Level-7'

AnWe11 committed:
Add time and update storage syncing issues

AnWe11 committed:
Merge branch 'branch-Level-8'

AnWe11 committed:
Add UI class

AnWe11 committed:
Package everything

AnWe11 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

AnWe11 committed:
Add JUnit for UI, Todo and Deadline tasks

AnWe11 committed:
Make Jar File

AnWe11 committed:
Merge branch 'branch-A-Jar'

AnWe11 committed:
Add comments

AnWe11 committed:
Merge branch 'branch-A-JavaDoc'

AnWe11 committed:
Add find function

AnWe11 committed:
Merge branch 'branch-Level-9'

AnWe11 committed:
Reformat comments

AnWe11 committed:
Add GUI

AnWe11 committed:
Merge branch 'branch-Level-10'

AnWe11 committed:
Tidy up comments

chanjunlin commented on src/main/java/commands/ByeCase.java:
```diff
@@ -0,0 +1,18 @@
+
+    public ByeCase() { }
+
+    /**
+     * Returns goodbye message and exit the program.
+     */
```
Nice use of Javadoc comments for the actions() method! it clearly states what the method does. You might want to specify that this method returns a string when called. 😊

chanjunlin commented on src/main/java/commands/CommandsParser.java:
```diff
@@ -0,0 +1,92 @@
+
+
+    // List of command keywords the user can input that will trigger a method call.
+    enum Keywords {
+        BYE, LIST, MARK, UNMARK, TODO, DEADLINE, EVENT, DELETE, FIND, INVALID;
```
I like how you used enum to manage command keywords. It makes your code cleaner and easier to manage! Would it be better to handle INVALID as part of error checking within the parsing logic 🤔?

chanjunlin commented on src/main/java/commands/CommandsParser.java:
```diff
@@ -0,0 +1,92 @@
+                responseString = todoCase.action();
+                break;
+            case DEADLINE:
+                DeadlineCase deadlineCase = new DeadlineCase(input, taskManager, data);
+                responseString = deadlineCase.action();
+                break;
+            case EVENT:
+                EventsCase eventCase = new EventsCase(input, taskManager, data);
+                responseString = eventCase.action();
+                break;
+            case LIST:
+                ListCase listCase = new ListCase(taskManager);
+                responseString = listCase.action();
+                break;
+            case MARK:
+            case UNMARK:
+                MarkUnmarkCase markUnmarkCase = new MarkUnmarkCase(input, keyword, taskManager, data);
+                responseString = markUnmarkCase.action();
+                break;
+            case BYE:
+                ByeCase byeCase = new ByeCase();
+                responseString = byeCase.action();
+                break;
+            case DELETE:
+                DeleteCase deleteCase = new DeleteCase(input, taskManager, data);
+                responseString = deleteCase.action();
+                break;
+            case FIND:
+                FindCase findCase = new FindCase(input, taskManager);
+                responseString = findCase.action();
+                break;
+            case INVALID:
+                InvalidCase invalidCase = new InvalidCase(input);
+                responseString = invalidCase.action();
+                break;
+            }
+        } catch (InvalidCommandException e) {
+            System.out.println(e.getMessage());
+            responseString = e.getMessage();
+        }
```
The use of switch based on the keyword enum makes it easy to follow which command maps to which actions. You Perhaps adding a default case in your switch statement can handle any unforeseen scenarios gracefully?

chanjunlin commented on src/main/java/commands/EventsCase.java:
```diff
@@ -0,0 +1,67 @@
+//            System.out.println("_____________________________________________");
+//            System.out.println("Input format for events tasks should be\nevent <Task Description> /from <from date> /to <to date>");
+//            System.out.println("_____________________________________________");
+            taskDescription = "";
```
I think this assignment might be unnecessary as you are not using taskDescription afterwards?

chanjunlin commented on src/main/java/commands/FindCase.java:
```diff
@@ -0,0 +1,65 @@
+                }
+            }
+
+            System.out.println("_____________________________________________");
```
I noticed that this string is typed out multiple times across different files. Would it be beneficial to define it as a final constant in your UI file? This way you can reference wherever needed and making future updates might be easier 🤔?

chanjunlin commented on src/main/java/exceptions/InvalidCommandException.java:
```diff
@@ -0,0 +1,13 @@
+package exceptions;
+
+public class InvalidCommandException extends Exception {
```
Maybe adding Javadoc comments to describe the class's purpose and usage? This could help people understand when this exception is thrown 🤔?

chanjunlin reviewed (commented):
LGTM. Good work on the Luigi implementation! The structure is clear and i appreciate your attention to error handling as well 😄

AnWe11 committed:
Delete commented code out for cleaner code.

Delete command: fix delete case.

Add welcome message to when the user launches Luigi Chatbot.

AnWe11 committed:
Add assertions to code.

Specifically when user inputs a negative taskID or a taskID that
is does not exist (bigger than total number of tasks).

Having these assertions will improve code quality and will ensure
that command cases such as Mark, Unmark and Delete will work.
Previously, it was handled as an exception but now, it is
handled as an assertion.

AnWe11 committed:
Merge pull request #1 from AnWe11/branch-A-Assertions

Add assertions

AnWe11 committed:
Add a new class StringChecker to handle if task description is
valid.

Previously, there were a lot of duplicated code and it in different
classes. I abstracted the code out and placed it in a static class
so that the code is cleaner and easier to read.

Following this, if there are any duplicated code in more classes
that requires checks, I can simply add them to the StringChecker
to practice good OOP.

AnWe11 committed:
Merge branch 'master' into branch-A-CodeQuality

AnWe11 committed:
Merge pull request #2 from AnWe11/branch-A-CodeQuality

Add StringChecker class

AnWe11 committed:
Refactored code in TaskManager class to include the use of streams
to make code more readable and functional.

Having streams allow for more readable code and also has the
flexibility of changing it to be a lazy stream for efficiency.

AnWe11 committed:
Merge branch 'master' into branch-A-Streams

AnWe11 committed:
Merge pull request #3 from AnWe11/branch-A-Streams

Add Streams

AnWe11 committed:
Clean up code

AnWe11 committed:
Create main.yml

AnWe11 committed:
Refactor code for Find case, want to include more features

AnWe11 committed:
Add more code

AnWe11 committed:
Refactor find feature to accept more search functions such as by
keyword, month, and type of task (todo, deadline or event).

Users can also use the find feature and type the specific filter
words in any order and the correct tasks will still be returned.

This will also make any future functions of find feature more
flexible to implement. The methods are abstracted out and each
individual filter has their own unique function.

AnWe11 committed:
Merge pull request #4 from AnWe11/branch-Extension

Branch extension

AnWe11 committed:
Update main.yml

AnWe11 committed:
Tweak GUI

AnWe11 committed:
Beautify GUI

AnWe11 committed:
Merge pull request #6 from AnWe11/branch-A-BetterGUI

Better GUI

AnWe11 committed:
Update ReadME

AnWe11 committed:
Edit ReadME
